### DISCUSSION TITLE
Adding float capability

### DISCUSSION THREAD

**Discussion by Russell-S-Harper** [cc65/cc65#2149] [Ideas]
I'd like to take a stab at adding `float` capability (single-precision floating-point format). I understand CC65 is undergoing a major rewrite, so what I propose is a light integration with no modifications of existing utilities required.

Will write numerous functions like:

```
float flt_add(float a, float b);
float flt_subtract(float a, float b);
float flt_multiply(float a, float b);
float flt_divide(float a, float b);
float flt_negated(float a);
float flt_reciprocal(float a);
```

As well as type conversion, and string input / output.

The intent is to support [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754) with normal, sub-normal, zero, NaN, and infinities.

The integration part will be along the lines of `typedef unsigned long float;` then rely on this build sequence:

- preprocess using `-E`
- parse the output for float expressions and convert them to function calls
- continue compiling to assembler, linking, etc.

For example:

```
float x = 7.0f;
float y = 6.0f;

float z = y * y - x * x;
```

Might be converted by the parser to something like:

```
unsigned long x = 0x40E00000;
unsigned long y = 0x40C00000;

unsigned long z = flt_subtract(flt_multiply(y, y), flt_multiply(x, x));
```

So two parts:

- float routines
- float parser

If that's okay with everyone, I can start. First draft will just demonstrate feasibility, for sure won't be size- or performance-optimized.

Subsequent drafts can add square root, log, trig, etc.

**Comment by mrdudz** [comment]
Such library exists (at least for CBM machines), see https://github.com/mrdudz/cc65-floatlib

I don't like the idea of putting extra magic preprocessing for this into the compiler/preprocessor - that will only create a big mess.

However, i have also already started adding support to the compiler - have a look at this ticket: https://github.com/cc65/cc65/pull/1777 (i see now that it doesn't merge cleanly anymore, i will fix this ASAP)

If you want to help pushing this forward, please do this via PR to this branch: https://github.com/mrdudz/cc65/tree/fptest

If you want to help with implementing the library, please use the API defined in this branch

**Reply by Russell-S-Harper** [reply]
Thank you for the links. Will review.

**Comment by mrdudz** [comment]
> i see now that it doesn't merge cleanly anymore, i will fix this ASAP

Done. Some of the new float tests fail... but thats expected :)

**Comment by Russell-S-Harper** [comment]
I see you and your team are making more progress in this project! I reviewed the code for the `cc65` compiler and other utiities, and unfortunately it would require too much investment on my part to be a useful contributor here. As a mental exercise, I did try writing some IEEE 754 compatible operations and functions, and with a bit of help from `gcc -fdiagnostics-format=json` was able to write a parser in PHP. I called it FLT, and a [repository](https://github.com/Russell-S-Harper/FLT) has just been created. While I think `cc65` is on a good path, users are welcome to try out FLT as preparation for when `cc65` is ready. Will close this ticket.

Best of luck!
